### PR TITLE
evaluate_batch with support of AMP

### PR DESCRIPTION
**Pull request by Adel-Moumen**
## What does this PR do?

@TParcollet found that `evaluate_batch` was lacking the support of AMP. This PR address this issue by adding the support of AMP and it also switch from `@torch.no_grad()` to `@torch.inference_mode()` (see: https://pytorch.org/docs/stable/generated/torch.inference_mode.html) `Code run under this mode gets better performance by disabling view tracking and version counter bumps.`

<details>
  <summary><b>Before submitting</b></summary>

- [ ] Did you read the [contributor guideline](https://speechbrain.readthedocs.io/en/latest/contributing.html)?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing [tests](https://github.com/speechbrain/speechbrain/tree/develop/tests) pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Does your code adhere to project-specific code style and conventions?

</details>

## PR review

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
- [ ] Confirm that the changes adhere to compatibility requirements (e.g., Python version, platform)
- [ ] Review the self-review checklist to ensure the code is ready for review

</details>

### CODE REVIEW

File: speechbrain/core.py [hunk 1]
```diff
@@ -1273,7 +1273,7 @@ def on_fit_batch_end(self, batch, outputs, loss, should_step):
         """
         pass
 
-    @torch.no_grad()
+    @torch.inference_mode()
```
Comment (TParcollet) [inline speechbrain/core.py:1276]: @Adel-Moumen this breaks backward compatibility with older version of torch no?

Comment (Adel-Moumen) [reply speechbrain/core.py:1276]: `@torch.inference_mode()` came in v1.9. If you want I can swap it with `@torch.no_grad()`.

Comment (TParcollet) [reply speechbrain/core.py:1276]: Yep swap it.

Comment (Adel-Moumen) [reply speechbrain/core.py]: Done.

**Review by TParcollet — APPROVED** [review]
LGTM
